### PR TITLE
init version that configures the all sensors for home assistant

### PR DESCRIPTION
**Pull request by marcelloceschia**
Hi @kbialek I have created a plugin that relates to #117 and can automatically configure all the sensors for home assistant.
Not sure if you like the changes I did in sensors and for the mqtt client - just let me know if i should change it

### CODE REVIEW

File: src/deye_sensors_igen_dtsd422.py [hunk 1]
```diff
@@ -54,15 +54,16 @@
...
+    unit="W", device_class='power',
     groups=["igen_dtsd422"],
 )
 ct1_reactive_power_sensor = SignedMagnitudeDoubleRegisterSensor(
     "Reactive Power CT1",
     0x17,
     1,
     mqtt_topic_suffix="ct1/reactive_power",
-    unit="Var",
+    unit="var",
```
Comment (kbialek) [inline src/deye_sensors_igen_dtsd422.py:65]: Is this change required by HA?

File: src/deye_mqtt.py [hunk 2]
```diff
@@ -35,48 +35,48 @@ def __init__(self, message: str):
 class DeyeMqttClient:
     def __init__(self, config: DeyeConfig):
         self.__log = logging.getLogger(DeyeMqttClient.__name__)
-        self.__mqtt_client = paho.Client(
+        self._mqtt_client = paho.Client(
```
Comment (kbialek) [inline src/deye_mqtt.py:38]: Please restore the original name. In this project private fields are prefixed with double underscore.

File: plugins/deye_plugin_ha-autoconfig.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+from deye_config import DeyeConfig
+from deye_mqtt import DeyeMqttClient, DeyeMqttPublishError
+from deye_plugin_loader import DeyePluginContext
+from deye_events import DeyeEventProcessor, DeyeEvent, DeyeObservationEvent
+
+from deye_sensor import SingleRegisterSensor
+from deye_sensors import sensor_list
+
+    
+class MqttAutoConfPublisher(DeyeMqttClient):
```
Comment (kbialek) [inline plugins/deye_plugin_ha-autoconfig.py:27]: Why is it necessary to extend `DeyeMqttClient` class?

File: plugins/deye_plugin_ha-autoconfig.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        """
+        
+
+        self.config = plugin_context.config
+        self.sensors = sensor_list
+        self.mqtt_client = MqttAutoConfPublisher(plugin_context.config)
+        self.mqtt_client.connect()
+        for sensor in self.sensors:
+            self.__configure_sensors(sensor)
+        self.mqtt_client.disconnect()
```
Comment (kbialek) [inline plugins/deye_plugin_ha-autoconfig.py:67]: If you disconnect here, then the event publisher will have to connect again right away. Imo, you can leave the connection up.

**Review by kbialek — CHANGES_REQUESTED** [review]
Thank you for submitting this PR. It's great to see, that you found the plugins mechanism useful here.

Please enhance the documentation. Ideally add a dedicated README section `Integration with Home-Assistant` after the `Installation` section.

File: src/deye_sensors_igen_dtsd422.py (hunk 1 above)
Comment (marcelloceschia) [reply src/deye_sensors_igen_dtsd422.py:65]: Yes, and it is the si unit for reactive power

File: src/deye_mqtt.py (hunk 2 above)
Comment (marcelloceschia) [reply src/deye_mqtt.py:38]: I like to re-use the implementation in `MqttAutoConfPublisher` as it needs to publish the configuration for sensors via mqtt

File: plugins/deye_plugin_ha-autoconfig.py (hunk 4 above)
Comment (marcelloceschia) [reply plugins/deye_plugin_ha-autoconfig.py:67]: This is a 2nd client that only publish the configuration, no need to keep this connection open. And i also saw issues when keeping this open

File: plugins/deye_plugin_ha-autoconfig.py (hunk 3 above)
Comment (marcelloceschia) [reply plugins/deye_plugin_ha-autoconfig.py:27]: for publishing the config `publish_sensor`

**Comment by rosch100** [comment]
I checked out the branch `marcelloceschia:plugin/ha-autoconfig`
and got this error in the plugin:
```
2023-08-19 16:25:43 2023-08-19 14:25:43,144 - DeyeMqttClient - INFO - Successfully connected to MQTT Broker located at mqtt.server.local:1883
2023-08-19 16:25:43   File "/opt/deye_inverter_mqtt/./deye_docker_entrypoint.py", line 59, in <module>
2023-08-19 16:25:43     main()
2023-08-19 16:25:43   File "/opt/deye_inverter_mqtt/./deye_docker_entrypoint.py", line 55, in main
2023-08-19 16:25:43     daemon_main()
2023-08-19 16:25:43   File "/opt/deye_inverter_mqtt/deye_daemon.py", line 142, in main
2023-08-19 16:25:43     daemon = DeyeDaemon(config)
2023-08-19 16:25:43   File "/opt/deye_inverter_mqtt/deye_daemon.py", line 57, in __init__
2023-08-19 16:25:43     plugin_loader.load_plugins(plugin_context)
2023-08-19 16:25:43   File "/opt/deye_inverter_mqtt/deye_plugin_loader.py", line 55, in load_plugins
2023-08-19 16:25:43     self.__plugins.append(plugin_module.DeyePlugin(plugin_context))
2023-08-19 16:25:43   File "/opt/deye_inverter_mqtt/plugins/deye_plugin_ha-autoconfig.py", line 66, in __init__
2023-08-19 16:25:43     self.__configure_sensors(sensor)
2023-08-19 16:25:43   File "/opt/deye_inverter_mqtt/plugins/deye_plugin_ha-autoconfig.py", line 80, in __configure_sensors
2023-08-19 16:25:43     'device_class': f"{sensor.device_class}",
2023-08-19 16:25:43 AttributeError: 'SingleRegisterSensor' object has no attribute 'device_class'
```

File: plugins/deye_plugin_ha-autoconfig.py (hunk 4 above)
Comment (kbialek) [reply plugins/deye_plugin_ha-autoconfig.py:67]: > And i also saw issues when keeping this open

Perhaps, because both connections use this same `mqtt client id`?

Why is it necessary to use two mqtt connections? For me it does not seem to be right.

Comment (marcelloceschia) [reply plugins/deye_plugin_ha-autoconfig.py:67]: Its not necessary to use two connections, I can change this

**Comment by rosch100** [comment]
> can you check this line https://github.com/marcelloceschia/deye-inverter-mqtt/blob/f1c361ea9fa40b66c1326ecf46fb600a5ddda705/src/deye_sensor.py#L77 is it the same in your code?

It's the same. Checked out recent changes (from today). Same error message.

**Comment by marcelloceschia** [comment]
You are using the original image that does not contain the changes I need to do for sensor type

**Comment by kbialek** [comment]
@marcelloceschia I think this feature is worth to be included in the service itself. WDYT?

**Comment by marcelloceschia** [comment]
@kbialek I'm not sure about that. I think with a plugin it is okay, because not all users needs it.
From plugin side, I only need the option to send a custom mqtt command for publishing the discovery information. I like your plugin concept, so why not stay with that?

**Comment by kbialek** [comment]
@marcelloceschia I'm assessing the following aspects:
1. Feature code integrity - your plugin code is not self-contained, it requires changes in the service code too, e.g. `device_class` sensor property is exclusively used by your plugin.
2. Feature scope consistency - the goal of this service is to publish metrics over mqtt, your plugin contributes to that goal by making it easier for people to integrate it.

Considering these two facts, I see, that this feature fits into the service concept, so I'm proposing to turn this plugin into a service feature. This service already has more optional features in place, so adding this one shouldn't be a problem.

**Comment by ixs** [comment]
This is an interesting PR. Nice idea. I like it.
Asking out of curiosity, not trying to imply that this kind of change should not be part of the `deye-inverter-mqtt`: Homeassist already has an integration at https://github.com/StephanJoubert/home_assistant_solarman which should be covering nearly every device this code base here covers. What use-case did you had in mind for this PR?

A thing I do not like is the addition of the `device_class` parameter for every sensor value all over the place.
Would it not be easier to have a mapping table that defines e.g. the unit `W`(att) to be an `energy` device type and just add the appropriate type when exporting to homeassistant?

**Comment by DivusJulius** [comment]
@ixs One reason to use this over the Solarman Integration is that the Solarman Integration is a custom integration. Not every user wants to run custom integrations. Getting the data over mqtt is a good alternative. 

Also the Solarman Integration just did not work for me and my M60/80/100G3-EU-Q0 :)

**Comment by bhuebschen** [comment]
any news here?

**Comment by Kolbi** [comment]
Still would like to see a easy solution for integrating it into Home-Assistant either Plugin or directly like this PR.

**Comment by kbialek** [comment]
HA support is added in https://github.com/kbialek/deye-inverter-mqtt/pull/194
